qdax/core/emitters/mees_emitter.py

```diff
@@ -0,0 +1,613 @@
+                ),
+                parent,
+            )
+            # Noise used for gradient computation
+            gradient_noise = sample_noise
+
+        # Expanding dimension to number of samples
+        samples = jax.tree_map(
+            lambda x: jnp.repeat(x, total_sample_number, axis=0),
+            parent,
+        )
+
+        # Applying noise to each sample
+        samples = jax.tree_map(
+            lambda mean, noise: mean + self._config.sample_sigma * noise,
+            samples,
+            sample_noise,
+        )
+
+        ######################
+        # Evaluating samples #
+
+        fitnesses, descriptors, extra_scores, random_key = self._scoring_fn(
+            samples, random_key
+        )
+
+        #######################
+        # Computing gradients #
+
+        # Choosing the score to use for rank
+        scores = jax.lax.cond(
+            (generation_count // self._config.num_optimizer_steps) % 2 == 0,
+            lambda scoring: self._novelty(
+                generation_count + 1, scoring[1], emitter_state.novelty_archive
+            ),
+            lambda scoring: scoring[0],
+            (fitnesses, descriptors),
+        )
+
+        # Computing rank with normalisation
+        if self._config.sample_rank_norm:
+
+            # Ranking objective
+            ranking_indices = jnp.argsort(scores, axis=0)
+            ranks = jnp.argsort(ranking_indices, axis=0)
+
+            # Normalising ranks to [-0.5, 0.5]
+            ranks = (ranks / (total_sample_number - 1)) - 0.5
+
+        # Computing rank without normalisation
+        else:
+            ranks = scores
+
+        # Reshaping rank to match shape of genotype_noise
+        if self._config.sample_mirror:
+            ranks = jnp.reshape(ranks, (sample_number, 2))
+            ranks = jnp.apply_along_axis(lambda rank: rank[0] - rank[1], 1, ranks)
+
+        ranks = jax.tree_map(
+            lambda x: jnp.reshape(
+                jnp.repeat(ranks.ravel(), x[0].ravel().shape[0], axis=0), x.shape
+            ),
+            gradient_noise,
+        )
+
+        # Computing the gradients
+        gradient = jax.tree_map(
+            lambda noise, rank: jnp.multiply(noise, rank),
+            gradient_noise,
+            ranks,
+        )
+        gradient = jax.tree_map(
+            lambda x: jnp.reshape(x, (sample_number, -1)),
+            gradient,
+        )
+        gradient = jax.tree_map(
+            lambda g, p: jnp.reshape(
+                -jnp.sum(g, axis=0) / (total_sample_number * self._config.sample_sigma),
+                p.shape,
+            ),
+            gradient,
+            parent,
+        )
+
+        # Adding regularisation
+        gradient = jax.tree_map(
+            lambda g, p: g + self._config.l2_coefficient * p,
+            gradient,
+            parent,
+        )
+
+        ######################
+        # Applying gradients #
+
```

Comment:
It would be nice to avoid having this case distinction here (cf. my comments about the optimizers)

Reply:
Modified as well with previous comment.